extensions/ql-vscode/src/remote-queries/variant-analysis-results-manager.ts

```diff
@@ -0,0 +1,178 @@
+import * as fs from 'fs-extra';
+import * as os from 'os';
+import * as path from 'path';
+
+import { Credentials } from '../authentication';
+import { Logger } from '../logging';
+import { AnalysisAlert, AnalysisRawResults } from './shared/analysis-result';
+import { sarifParser } from '../sarif-parser';
+import { extractAnalysisAlerts } from './sarif-processing';
+import { CodeQLCliServer } from '../cli';
+import { extractRawResults } from './bqrs-processing';
+import { VariantAnalysisScannedRepositoryResult } from './shared/variant-analysis';
+import { DisposableObject, DisposeHandler } from '../pure/disposable-object';
+import { VariantAnalysisRepoTask } from './gh-api/variant-analysis';
+import * as ghApiClient from './gh-api/gh-api-client';
+import { EventEmitter } from 'vscode';
+
+type CacheKey = `${number}/${string}`;
+
+const createCacheKey = (variantAnalysisId: number, repoTask: VariantAnalysisRepoTask): CacheKey => `${variantAnalysisId}/${repoTask.repository.full_name}`;
+
+export type ResultDownloadedEvent = {
+  variantAnalysisId: number;
+  repoTask: VariantAnalysisRepoTask;
+}
+
+export class VariantAnalysisResultsManager extends DisposableObject {
+  private readonly cachedResults: Map<CacheKey, VariantAnalysisScannedRepositoryResult>;
+
+  private readonly _onResultDownloaded = this.push(new EventEmitter<ResultDownloadedEvent>());
+  readonly onResultDownloaded = this._onResultDownloaded.event;
+
+  private readonly _onResultLoaded = this.push(new EventEmitter<VariantAnalysisScannedRepositoryResult>());
+  readonly onResultLoaded = this._onResultLoaded.event;
+
+  constructor(
+    private readonly cliServer: CodeQLCliServer,
+    private readonly storagePath: string,
+    private readonly logger: Logger,
+  ) {
+    super();
+    this.cachedResults = new Map();
+  }
+
+  public async download(
+    credentials: Credentials,
+    variantAnalysisId: number,
+    repoTask: VariantAnalysisRepoTask,
+  ): Promise<void> {
+    if (!repoTask.artifact_url) {
+      throw new Error('Missing artifact URL');
+    }
+
+    const resultDirectory = this.getRepoStorageDirectory(variantAnalysisId, repoTask.repository.full_name);
+
+    const result = await ghApiClient.getVariantAnalysisRepoResult(
+      credentials,
+      repoTask.artifact_url
+    );
+
+    fs.mkdirSync(resultDirectory, { recursive: true });
+    await fs.writeFile(path.join(resultDirectory, 'results.zip'), JSON.stringify(result, null, 2), 'utf8');
+
+    this._onResultDownloaded.fire({
+      variantAnalysisId,
+      repoTask,
+    });
+  }
+
+  public async loadResults(
+    variantAnalysisId: number,
+    repoTask: VariantAnalysisRepoTask
+  ): Promise<VariantAnalysisScannedRepositoryResult> {
+    const result = this.cachedResults.get(createCacheKey(variantAnalysisId, repoTask));
+    if (result) {
+      return result;
+    }
+
+    return this.loadResultsIntoMemory(variantAnalysisId, repoTask);
```

Comment:
You could shorten this using the [`??` operator](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Nullish_coalescing_operator):

```
return result ?? this.loadResultsIntoMemory(variantAnalysisId, repoTask);
```